libkineto/src/ActivityProfiler.cpp

```diff
@@ -186,6 +190,37 @@ void ActivityProfiler::processCpuTrace(
 
   CpuGpuSpanPair& span_pair = recordTraceSpan(cpuTrace.span, cpuTrace.gpuOpCount);
   TraceSpan& cpu_span = span_pair.first;
+  // Update act.unique_op_id to correlate forward-op and backward-op.
+  for (auto & act : cpuTrace.activities) {
+    if (act.fwdThreadId > 0) {
+      // act is backward op.
+      uint64_t key = GET_FORWARD_THREAD_KEY(act.fwdThreadId, act.seqNumber);
+      auto iter = tidSeq2activity_.find(key);
+      if (iter != tidSeq2activity_.end()) {
+        act.unique_op_id = iter->second->unique_op_id;
+      }
+    }
+    else if (act.startThreadId != 0 && act.seqNumber >= 0) {
+      // act is forward op.
+      uint64_t key = GET_FORWARD_THREAD_KEY(act.startThreadId, act.seqNumber);
+      // Assumption: Among all ops with same sequence number,
+      // the one with biggest start time is most likely launching backward op.
+      auto iter = tidSeq2activity_.find(key);
+      if (iter == tidSeq2activity_.end()) {
+        uint64_t uid = tidSeq2activity_.size() + 1; // Start from 1.
+        act.unique_op_id = uid;
+        tidSeq2activity_[key] = &act;
+      }
+      else {
+        if (act.startTime >= iter->second->startTime) {
```

Comment:
When does it happen that the same key is encountered multiple times?

Reply:
Now the sequence number is only incremented on creating a "Node" object for backward pass, by [calling "at::sequence_number::get_and_increment()"](https://github.com/pytorch/pytorch/blob/bf7d03ff1f3cded53a16b707a2db76ce25c130f7/torch/csrc/autograd/function.h#L130). 
Among all ops with same sequence number, the one with biggest startTime is the one launching backward op. So I choose this one to cache into "tidSeq2activity_" for matching with backward-op of same tid and seq number.
I've added comments explaining this.